extensions/wikia/CategoryPagination/CategoryPaginationViewer.class.php

```diff
@@ -2,102 +2,221 @@
-			return $maxPaginator->getHeadItem();
-		}
-		return '';
+			return false;
+		}, function () {
+			return $this->formatList( $this->articles, $this->articles_start_char );
+		} );
+
+		// Get media
+		$this->processSection( self::TYPE_FILE, function ( $title, $humanSortkey, $row ) {
+			$this->addImage( $title, $humanSortkey, $row->page_len, $row->page_is_redirect );
+			return true;
+		}, function () {
+			if ( $this->showGallery ) {
+				return $this->gallery->toHTML();
+			}
+			return $this->formatList( $this->imgsNoGallery, $this->imgsNoGallery_start_char );
+		} );
+
+		// Get subcategories
+		$this->processSection( self::TYPE_SUBCAT, function ( $title, $humanSortkey, $row ) {
+			$cat = Category::newFromTitle( $title );
+			if ( $this->tree ) {
+				$this->children[] = $this->tree->renderNodeInfo( $title, $cat );
+				$this->children_start_char[] = $this->getSubcategorySortChar( $title, $humanSortkey );
+			} else {
+				$this->addSubcategoryObject( $cat, $humanSortkey, $row->page_len );
+			}
+			return true;
+		}, function () {
+			return $this->formatList( $this->children, $this->children_start_char );
+		} );
 	}
 
 	/**
-	 * Select the first page matching given type and page number and return the human-readable
-	 * sort key associated with it
+	 * Process one section. Updates $this->numberShown with number of added items
+	 * If $processRow callable returns false for a given row it's not counted.
+	 * Updates $this->content with the value returned by $generateContent callable.
 	 *
-	 * This value is passed to the original CategoryViewer code as the $this->from member.
-	 * The DB is called, but the query uses the proper index and therefor is super fast (< 10ms),
-	 * so there's no need to cache.
+	 * @param string $type one of self::TYPE_*
+	 * @param callable $processRow run this function for each of the rows from the DB
+	 *                    the following params are passed to it: ( $title, $humanSortkey, $row )
+	 * @param callable $generateContent run this function if number of rows added is non-zero.
+	 *                    the returned value is saved to $this->content. No params are passed to it
 	 */
-	private function getActualFrom( $type, $page ) {
-		$dbr = wfGetDB( DB_SLAVE, 'category' );
-
-		// if beyond the last page, show the last page instead
-		$totalPages = ceil( $this->counts[$type] / $this->limit );
-		if ( $page > $totalPages && $totalPages >= 1 ) {
-			$page = $totalPages;
-		}
-
-		$conditions = [
-			'cl_to' => $this->title->getDBkey(),
+	private function processSection( string $type, callable $processRow, callable $generateContent ) {
+		$conds = [
 			'cl_type' => $type,
+			'cl_to' => $this->title->getDBkey(),
 		];
 
-		$res = $dbr->select(
+		// Wall uses this hook to filter out wall messages from the category pages
+		wfRunHooks( 'CategoryViewer::beforeCategoryData', array( &$conds ) );
+
+		$dbr = wfGetDB( DB_SLAVE, 'category' );
+		$result = $dbr->select(
 			[ 'page', 'categorylinks' ],
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "categorylinks" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1192)